src/osd/ReplicatedBackend.cc

```diff
@@ -1698,22 +1719,68 @@ void ReplicatedBackend::submit_push_data(
   }
 
   if (first) {
-    t->remove(coll, ghobject_t(target_oid));
-    t->touch(coll, ghobject_t(target_oid));
+    if (!complete) {
+      t->remove(coll, ghobject_t(target_oid));
+      t->touch(coll, ghobject_t(target_oid));
+      bufferlist bv = attrs[OI_ATTR];
+      object_info_t oi(bv);
+      t->set_alloc_hint(coll, ghobject_t(target_oid),
+			oi.expected_object_size,
+			oi.expected_write_size,
+			oi.alloc_hint_flags);
+    } else {
+      //remove xattr and update later if overwrite on original object
+      t->rmattrs(coll, ghobject_t(target_oid));
+      //if need update omap, clear the previous content first
+      if (clear_omap)
+        t->omap_clear(coll, ghobject_t(target_oid));
+    }
     t->truncate(coll, ghobject_t(target_oid), recovery_info.size);
     t->omap_setheader(coll, ghobject_t(target_oid), omap_header);
 
-    bufferlist bv = attrs[OI_ATTR];
-    object_info_t oi(bv);
-    t->set_alloc_hint(coll, ghobject_t(target_oid),
-		      oi.expected_object_size,
-		      oi.expected_write_size,
-		      oi.alloc_hint_flags);
+    if (!complete) {
+      //clone overlap content in local object
+      struct stat st;
+      int r = store->stat(coll, ghobject_t(recovery_info.soid), &st);
+      if (recovery_info.object_exist) {
+	assert(r == 0);
+        uint64_t local_size = MIN(recovery_info.size, (uint64_t)st.st_size);
```

Comment:
It can, but in order to resolve the issue (if the remote indicates {0~1024, 10240~1024}, which need to recovery. The hole exists {8192~1024} while the local object size is 8192, so we need to skip the local clone for the holes). In this way, we can deal with it.  Another option is that calculating clone_interval on the remote side and transmit it inside recovery_info. I think you prefer the later one, right? I will modify it